### ISSUE TITLE
`ugrep -Q ~` can freeze while trying to edit settings

### ISSUE THREAD

**Issue by stdedos** [Genivia/ugrep#90]
On a sufficiently large `~`:

* `ugrep -Q ~`
* Type a pattern
* Wait for _a match_
* Go to settings e.g. to disable binary matching

Expected: `ugrep` help / options panel is responsive
Actual: `ugrep` is frozen

Expected: `ugrep` responds to Ctrl+C / Ctrl+\
Actual: `ugrep` is frozen

**Comment by genivia-inc** [comment]
This needs a bit of clarification. Normal testing does not show an issue. Do you mean switching to the help screen when the results do not show up yet? Some options immediately affect the search, which will restart (even when incomplete), while some other options don't, so I'm curious if this happens in your case with the binary match option (only)?

**Comment by genivia-inc** [comment]
Also, what machine and what OS are you running on?

With a very large file system search like from home or from root, we do not see an issue with our test machines.

**Comment by stdedos** [comment]
> This needs a bit of clarification. Normal testing does not show an issue. Do you mean switching to the help screen when the results do not show up yet?

You forgot to activate "ignore binary files", and you only realized it when the first binary result comes in. You "scramble" to disable it, you switch to help/options menu, you try to find and press the shortcut, only to realize that `ugrep` is now not responding (and at 60%-80% CPU time)

> Some options immediately affect the search, which will restart (even when incomplete), while some other options don't, so I'm curious if this happens in your case with the binary match option (only)?

That was my usecase, as my go-to options are `grep -TrinIP (pattern) ~` (when directory is ~), and I had started `ugrep` just with `-Q`

> Also, what machine and what OS are you running on?

```
            .-/+oossssoo+/-.               
        `:+ssssssssssssssssss+:`           ------------------------------ 
      -+ssssssssssssssssssyyssss+-         OS: Ubuntu 18.04.5 LTS x86_64 
    .ossssssssssssssssssdMMMNysssso.       Host:
   /ssssssssssshdmmNNmmyNMMMMhssssss/      Kernel: 5.4.0-58-generic 
  +ssssssssshmydMMMMMMMNddddyssssssss+     Uptime: 10 days, 12 hours, 4 mins 
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/    Packages: 69569 (apt), 17 (snap) 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Shell: zsh 5.4.2 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   Resolution: 1920x1080 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   DE: GNOME 3.28.4 
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   WM: Mutter 
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   WM Theme: Adwaita 
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Theme: Ambiance [GTK2/3] 
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/    Icons: ubuntu-mono-dark [GTK2/3] 
  +sssssssssdmydMMMMMMMMddddyssssssss+     Terminal: gnome-terminal 
   /ssssssssssshdmNNNNmyNMMMMhssssss/      CPU: Intel i7-4600U (4) @ 3.300GHz 
    .ossssssssssssssssssdMMMNysssso.       GPU: Intel Haswell-ULT 
      -+sssssssssssssssssyyyssss+-         Memory: 13642MiB / 15914MiB 
        `:+ssssssssssssssssss+:`
            .-/+oossssoo+/-.     
```

> With a very large file system search like from home or from root, we do not see an issue with our test machines.

My home statistics (from qdirstat):
![image](https://user-images.githubusercontent.com/133706/103270489-b8ecb080-49c0-11eb-8a1e-044a40f69884.png)

**Comment by genivia-inc** [comment]
This information does not help. For optimization ugrep `-Q` never searches further than the results shown on screen plus another page. So it will not visit all files in the file system and certainly won't do that if you don't have a search pattern specified. Only if you have a pattern that no file matches, then the search continues. But that is not what you described.

We test on several systems, including searching from root and haven't found any problems on MacOS, Linux, Android, Cygwin, and Windows.

> You forgot to activate "ignore binary files", and you only realized it when the first binary result comes in. You "scramble" to disable it, you switch to help/options menu, you try to find and press the shortcut, only to realize that ugrep is now not responding (and at 60%-80% CPU time)

Switching options in the help menu while the search happens has no detrimental effect and is the same as switching options in the main screen with `ALT-I` for example. When an option is changed, an new search starts with the same pattern specified.

Since the CPU still runs the search continues. It is not clear what you mean by "freezing".

You may want to try option `-J1` to force one thread and see what happens.

**Comment by stdedos** [comment]
While I haven't tried your suggestion, I can tell you what happened when I left it running:

Same old, I ran something, I left it scanning, and then "after a while" I `Alt+I`'ed.
Again, what happened is that it froze (with `Searching.`), but I didn't try to kill it - I just left it there.

The result of this was; it finished searching the 111.8G file (that P99..P100 segment), and it returned a result from it. After that, everything continued normally (The TUI thread updating the status as expected).

I think you need to safely kill the searching threadpool pre-emptively; regardless if you have an active thread searching on a ginormous file or not.

**Comment by stdedos** [comment]
I also think that you don't freeze your searching threadpool the first time the user presses `Ctrl-C` (After I wrote the above comment, `ugrep` had finished searching, and produced two more results)

**Comment by stdedos** [comment]
Also also, if the user accidentally modifies the searched pattern, modifications to command line (e.g. `Alt+I`) are not preserved.

I don't know if that is only on my case or generally, but I think that this is deviating from user's expectations.
On `ugrep -Q`, I would spend some screen space (top line, ala `PS1`, or ala `PS4`) to keep track of the active settings. Invoked as `ugrep -TrinQ`:
* ```console
  Mod: initial tab, recurse, ignore case, line number
  Q> %query
  .
  .
  .
  ```
* ```console
  -Trin> %query
  .
  .
  .
  ```
* ```console
  Q> %query          <-Trin
  .
  .
  .
  ```

reflecting any future modifications

**Comment by genivia-inc** [comment]
> Also also, if the user accidentally modifies the searched pattern, modifications to command line (e.g. Alt+I) are not preserved.

Not true. The command line options are always preserved. Typing a pattern does not reset options. For example `-inTQ`:

<img width="570" alt="image" src="https://user-images.githubusercontent.com/13117347/103661927-49853c80-4f3d-11eb-9605-c73a53be346c.png">

The `Q>` prompt does not show all options. Only the search mode (`F`, `G`, `Z`, `Q`, `P`) is indicated. Options are shown in the help/option page, which is easier to understand that a long string of characters.

I don't know what ugrep version you have or what is changed to it, but this works fine.

**Comment by stdedos** [comment]
> Also also, if the user accidentally modifies the searched pattern, modifications to command line (e.g. `Alt+I`) are not preserved.

_[...] modifications to the `ugrep` options that deviate from the ones when invoked._

Apologies, wrong phrasing. I don't remember which one I am using, since I have updated it through this post. I'll check later and come back to you.

> The `Q>` prompt does not show all options. Only the search mode (`F`, `G`, `Z`, `Q`, `P`) is indicated. Options are shown in the help/option page, which is easier to understand that a long string of characters.

I know that; my proposal was to extend it. I am aware that users may not be accustomed to the flags (but that IMHO can be their problem 😕)

**Comment by genivia-inc** [comment]
> The result of this was; it finished searching the 111.8G file (that P99..P100 segment), and it returned a result from it. After that, everything continued normally (The TUI thread updating the status as expected).

This means that everything works as it should be. There is no "freezing" but a temporary delay due to several conditions that are not unexpected in this case.

A temporary slow down before a new search starts can happen when the following conditions are met:
1. the file being searched by a thread is very large, e.g. a couple of GB.
2. a search has started by a thread on this file and there are no matching patterns in the file.
3. the large file is not ignored from the search, e.g. ignored with `-I`.

The internal search engine is cancelled (to allow the next search to start) when a pattern match is found in the large file, because this allows the cancellation flag to be inspected by the thread.

I think these conditions are reasonable, because for those kind of searches it makes sense to start with `-I` to avoid delays when switching patterns and options when one or more large files are searched and those do not have any matches.

There are possible ways I can think of to stop a thread's internal search quicker, but that could negatively affect the search speed. For performance it is best not to mess around with the search algorithm.

**Comment by genivia-inc** [comment]
One thing we should do is replace "Searching..." with "Restarting" (or perhaps "Wait"?) to indicate that the search is restarting and temporarily waiting on all worker threads to end. This happens rarely and only when a searching thread has not yet completed a huge file, e.g. 100GB.

**Comment by stdedos** [comment]
> There are possible ways I can think of to stop a thread's internal search quicker, but that could negatively affect the search speed. For performance it is best not to mess around with the search algorithm.

* It makes no sense to gracefully wait on a thread whose results will be dumped anyway.
* I don't understand why a thread cannot be killed sooner.
* I don't know for how long time are we talking about (there is no quick `/usr/bin/time` for threads), but it's definitely not "a couple of seconds". It's more like a minute or more.
* "`main`" TUI thread also freezes. From GUI/UX perspective, "this is unacceptable".

> A temporary slow down before a new search starts can happen when the following conditions are met:

The temporary slowdown is not dependant on the No. 2 condition. My binary file had matches (one to be precise).

**Comment by genivia-inc** [comment]
The same thing can happen when you pipe grep for some large files through `less`, when no pattern matches the files. When `less` exits by typing `q`you may have to wait a bit long(er). The reason is that the `less` command closes the pipe but grep doesn't see that until it tries to output something and then it quits.

Makes more sense now?

**Comment by genivia-inc** [comment]
One other improvement for the upcoming release is a more thorough check for binary files, so that option `-I` is more effective to skip them. There are still cases when some binary files are not seen as binary until a search is conducted on them.

**Comment by stdedos** [comment]
Okay, if there is really nothing to be done - it's just that the delay "is a bit too much".

ugrep is so magical sometimes, I have trouble "accepting" weaknesses on it 😛

**Comment by genivia-inc** [comment]
I understand and I'm listening, but I'm not quite sure to call this a weakness. If the binary files you're (accidentally) including in your search are over 118GB and no pattern matches these files so they have to be searched from start to end, then some delays are to be expected. IMO it makes sense to enable `--binary-files=without-match` in your `.ugrep` configure. This means that `-ug -Q` will work faster.

**Comment by genivia-inc** [comment]
Besides the other minor changes that improve the user experience, the help screen should not restart a search until we exit the help screen. Initially I thought it would be beneficial to restart searches in the background whenever options are changed, running the searches behind the help screen so switching back gives the results already collected. However, if options are changed a lot and the search is temporarily delayed then this can affect the responsiveness of the help screen. Something to think about.

**Comment by genivia-inc** [comment]
I've made a couple of minor improvements to v3.1.2 to address the issues and resolutions suggested.

**Comment by genivia-inc** [comment]
I made a small mistake in the 3.1.2 update with `-Q` when switching back to the main screen from the help screen. That screen refresh problem is corrected now.
